clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1102,46 +1207,94 @@ public void subscribe(Pattern pattern, ConsumerRebalanceListener listener) {
+     * supported).
+     *
+     * @throws ConcurrentModificationException if another thread already has the lock
+     */
+    private void acquire() {
+        final Thread thread = Thread.currentThread();
+        final long threadId = thread.getId();
+        if (threadId != currentThread.get() && !currentThread.compareAndSet(NO_CURRENT_THREAD, threadId))
+            throw new ConcurrentModificationException("KafkaConsumer is not safe for multi-threaded access. " +
+                "currentThread(name: " + thread.getName() + ", id: " + threadId + ")" +
+                " otherThread(id: " + currentThread.get() + ")"
+            );
+        refCount.incrementAndGet();
+    }
+
+    /**
+     * Release the light lock protecting the consumer from multithreaded access.
+     */
+    private void release() {
+        if (refCount.decrementAndGet() == 0)
+            currentThread.set(NO_CURRENT_THREAD);
+    }
+
     private void subscribeInternal(Pattern pattern, Optional<ConsumerRebalanceListener> listener) {
-        maybeThrowInvalidGroupIdException();
-        if (pattern == null || pattern.toString().isEmpty())
-            throw new IllegalArgumentException("Topic pattern to subscribe to cannot be " + (pattern == null ?
+        acquireAndEnsureOpen();
+        try {
+            maybeThrowInvalidGroupIdException();
+            if (pattern == null || pattern.toString().isEmpty())
+                throw new IllegalArgumentException("Topic pattern to subscribe to cannot be " + (pattern == null ?
                     "null" : "empty"));
-
-        throwIfNoAssignorsConfigured();
-        log.info("Subscribed to pattern: '{}'", pattern);
-        subscriptions.subscribe(pattern, listener);
-        updatePatternSubscription(metadata.fetch());
-        metadata.requestUpdateForNewTopics();
+            throwIfNoAssignorsConfigured();
+            log.info("Subscribed to pattern: '{}'", pattern);
+            subscriptions.subscribe(pattern, listener);
+            updatePatternSubscription(metadata.fetch());
+            metadata.requestUpdateForNewTopics();
+        } finally {
+            release();
+        }
     }
 
     private void subscribeInternal(Collection<String> topics, Optional<ConsumerRebalanceListener> listener) {
-        maybeThrowInvalidGroupIdException();
-        if (topics == null)
-            throw new IllegalArgumentException("Topic collection to subscribe to cannot be null");
-        if (topics.isEmpty()) {
-            // treat subscribing to empty topic list as the same as unsubscribing
-            unsubscribe();
-        } else {
-            for (String topic : topics) {
-                if (isBlank(topic))
-                    throw new IllegalArgumentException("Topic collection to subscribe to cannot contain null or empty topic");
-            }
+        acquireAndEnsureOpen();
+        try {
+            maybeThrowInvalidGroupIdException();
```

Comment:
The original implementation drives me crazy. Apparently, the original implementation can both throw GroupId after or before acquiring the lock... depends on the API.  It seems like this is the right way to do this because our premise is that the consumer is not a thread-safe application so it makes sense to throw the lock exception before checking the groupId